Introduce the upgrade-config setting to control the behavior of image preloading during upgrades

starbops opened the pull request:
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**
<!-- Explain the problem you are aiming to resolve in this PR. -->

In the current Harvester upgrade implementation, image preloading will always happen (provided new container images are introduced) in a one-node-at-a-time fashion. It takes a significant portion of the upgrade time, no matter how powerful the given cluster nodes and network capacity. We shall provide the flexibility for users to customize the image preload strategy.

**Solution:**
<!-- Example: When "Adding a function to do X", explain why it is necessary to have a way to do X. -->

Add a new Setting called `upgrade-config`. The upgrade controller will honor the setting value to decide whether to skip preload, preload sequentially, or preload in parallel for container images. This new Setting object will be where we store upgrade-relevant configs/options. For example:

```json
{
  "imagePreloadOption": {
    "strategy": {
      "type": "parallel",
      "concurrency": 2
    }
  }
}
```

This will allow the Harvester cluster to conduct future upgrades with parallel image preloading at a two-node-at-a-time pace. If the `.imagePreloadOption.strategy.concurrency` field is set to 0 (this is also the default value), all nodes will start preloading images simultaneously. Concurrency value is ignored when a strategy type other than "parallel" is chosen.

Available `.imagePreloadOption.strategy.type` field are:

- `skip`
- `sequential`
- `parallel`

**Related Issue:**

#3059 #6028

**Test plan:**
<!-- Make sure tests pass on the Circle CI. -->

0. For reviewers, please use a custom-built ISO image that contains the PR; for QAs, please use the master ISO image.
1. Prepare a Harvester cluster consists of at lease two nodes
2. Configure the `upgrade-config` setting's `.imagePreloadOption.strategy.type` field to `skip`, `sequential`, or `parallel` respectively
   ```shell
   kubectl patch setting.harvesterhci upgrade-config --type=merge -p '{"value":"{\"imagePreloadOption\":{\"strategy\":{\"type\":\"skip\"}}}"}'
   kubectl patch setting.harvesterhci upgrade-config --type=merge -p '{"value":"{\"imagePreloadOption\":{\"strategy\":{\"type\":\"sequential\"}}}"}'
   kubectl patch setting.harvesterhci upgrade-config --type=merge -p '{"value":"{\"imagePreloadOption\":{\"strategy\":{\"type\":\"parallel\",\"concurrency\":2}}}"}'
   ```
4. Start the upgrade using the same ISO image
5. Observe the upgrade progress, when it enters the image-preload phase:
   - If `skip` is specified, there should be **no** `prepare` pods running in the cluster
   - If `sequential` is specified, there should be exact **one** `parepare` pod running at a time
   - If `parallel` is specified, there should be **two** (or other number depending on how many nodes are there) `prepare` pods running simultaneously
7. The upgrade ends successfully

starbops committed:
feat(setting): add new setting image-preload-strategy

Available values are skip, sequential, and parallel. The default value
is sequential.

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

starbops committed:
feat(upgrade): honor image-preload-strategy setting

When entering the image-preload phase, the controller will take the
corresponding action (to skip, preload sequentially, or in parallel)
according to the image-preload-strategy setting.

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

w13915984028 reviewed (commented):
A question about the parallel capacity.

brandboat reviewed (approved):
Thanks for the patch, LGTM !

bk201 reviewed (commented):
Thanks! Just leave some comments.

ibrokethecloud reviewed (approved):
lgtm. Do we want to bump up the upgrade repo vm specs from 1cpu/1G when parallel loading is requested?

starbops committed:
feat(upgrade): introduce upgrade-config setting

The upgrade-config setting will incorporate all Harvester
Upgrade-relevant options that users can fiddle with. The first option is
for the image preload phase and it's about configuring the image preload
strategy. Once set, the upgrade controller will follow the provided
option to do the preload, either skipping the phase totally, or
preloading images with specific concurrency.

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

w13915984028 reviewed (approved):
LGTM, thanks.

With a few suggestions.

brandboat reviewed (approved):
LGTM, thanks
